### PR TITLE
Comprehensively support Azure AD w/ ADFS

### PR DESCRIPTION
**Pull request by osulli**
## What

Users of ADFS in their Azure AD configuration will find that they must use `--mode=gui` in order to input Basic Authentication credentials into an ADFS prompt.

Before this PR, aws-azure-login did not support this potential scenario.

An attempt was made in #155, however insufficient progress and movement has occurred as well as the solution not fully covering potential ADFS configurations.

## Why

- This problem is notorious at my organisation for macOS & Linux users. All of the refresh and automation potential was ruined by the BasicAuth prompt from ADFS.

## Changelog

```text
# Changed
- Update README with `ADFS` section
- Fix minor typos / grammar
- Reduce the time aws-azure-login retries on `unrecognised states` from *30s* to *10s*
- Update unrecognised state error message with more clear suggestions and formatting

# Added
- Add `debug` and `lint` yarn commands
- Ignore `unrecognised-*` snapshots
- Support two new *optional* input variables: `adfs_prompt_expected` & `adfs_username`
- Add `Chrome exception` state which will capture Chromium and HTTP errors (e.g. 401 because credentials invalid)
```

### CODE REVIEW

**Comment by osulli** [comment]
@kennethaasan @olemarkus 
I've removed my support for #155 in favour of my own: #177 

If you could review it at your earliest convenience that would be much appreciated.

**Comment by kennethaasan** [comment]
Hi, and thanks for this PR @osulli!

To me, it looks like this feature should have its own states because it doesn't share any logic without ADFS (Correct me if I'm wrong!) That way you also avoid creating those global variables and the implementation will be easier to follow.

The states can be
 - username, password, and execute _page.authenticate_
 - Chrome exception

Does that make sense?

**Comment by osulli** [comment]
@kennethaasan 

Totally makes sense, and it was how I originally started. However, I was unable to find a suitable selector for the ADFS Basic Auth prompt. Upon inspection with dev tools, it has no tags and appears to almost be native? It is literally a 'non-inspectable' widget in terms of Chrome dev tools.
That in itself could very well be a criterion to test, but I felt if I was putting in a state that 'pushes' it onto anyone so I fell back to configuring an 'expectation' for it. I would be concerned that we said 'if the page is blank it must be ADFS' unless you as a maintainer waive that worry.

E: Saying that, this does look fairly decent at first glance (interception rather than detection via selection):
https://stackoverflow.com/a/64433034

**Comment by osulli** [comment]
Unfortunately, it looked like I never committed the other attempts, and waited until I had a working solution. However, I do recall a different approach involving prompting for the password prompt, awaiting a promise, right before awaiting the basic authentication prompt however this failed and was actually what led me to write the Chrome Exception handling.

**Comment by kennethaasan** [comment]
@osulli, I guess you can try to create a "dumb" selector for the Basic Auth step?

Please remove the version bump. I like to handle those myself in the main branch.

**Comment by osulli** [comment]
> @osulli, I guess you can try to create a "dumb" selector for the Basic Auth step?
Sorry, I have no idea how to go about doing that as I was already investigating what could be done. The only one is a 'if no selectors' but that is too 'catch all'. 

Do you have a suggested entrypoint?

> Please remove the version bump. I like to handle those myself in the main branch.

Affirmative!.. makes sense now you mention it.

File: README.md [hunk 1]
```diff
@@ -155,6 +155,39 @@ _Note:_ on Linux you will likely need to disable the Puppeteer sandbox or Chrome
...
+adfs_username=foo_bar
+```
+
+If you would like to fully automate your ADFS sign in, use the following configuration:
+
+```
+# ~/.aws/config
+[profile foo]
+azure_default_username=foo.bar@acme.com
+azure_default_password=sUp3RS3CR3tP422W0Rd!
```
Comment (mungojam) [inline README.md:186]: I've come to think that a general rule is that the config file shouldn't have any secrets in it, only the credentials file. That allows for people to have their config shared easily with others while never sharing their credentials file.

I might have made that rule up, but it feels like a good one. I guess you can't have arbitrary settings like this in the credentials file though?

**Pull request closed without merging**